<a href="https://colab.research.google.com/github/VaibhavRajput022/Codsoft/blob/main/TITANIC_SURVIVAL_PREDICTION.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Titanic Survival Prediction Model

## 1. Load Data

In [5]:
import pandas as pd

# Load the Titanic dataset from the provided path
try:
    df = pd.read_csv('/content/Titanic-Dataset.csv')
    print("Dataset loaded successfully.")
    display(df.head())
except FileNotFoundError:
    print("Error: The file 'Titanic-Dataset.csv' was not found. Please ensure it's in the correct directory.")
except Exception as e:
    print(f"An error occurred while loading the dataset: {e}")

Dataset loaded successfully.


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


## 2. Data Exploration

In [6]:
# Display basic information about the dataset, including data types and non-null values
print("\n--- Dataset Info ---")
df.info()

# Check for missing values in each column
print("\n--- Missing Values ---")
print(df.isnull().sum())

# Display basic statistical summary of numerical columns
print("\n--- Descriptive Statistics ---")
display(df.describe())


--- Dataset Info ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB

--- Missing Values ---
PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin   

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


## 3. Data Preprocessing

Based on the data exploration, we need to address missing values and convert categorical features into a numerical format. We will also drop columns that are not useful for prediction.

In [7]:
import numpy as np

# Drop columns that are unlikely to be useful for prediction or have too many missing values
df_processed = df.drop(['PassengerId', 'Name', 'Ticket', 'Cabin'], axis=1)

# Impute missing 'Age' values with the median age
df_processed['Age'].fillna(df_processed['Age'].median(), inplace=True)

# Impute missing 'Embarked' values with the most frequent value (mode)
df_processed['Embarked'].fillna(df_processed['Embarked'].mode()[0], inplace=True)

# Convert 'Sex' and 'Embarked' columns to numerical using one-hot encoding
df_processed = pd.get_dummies(df_processed, columns=['Sex', 'Embarked'], drop_first=True)

print("\n--- Dataset after Preprocessing ---")
display(df_processed.head())
print("\n--- Updated Dataset Info ---")
df_processed.info()
print("\n--- Missing Values after preprocessing ---")
print(df_processed.isnull().sum())


--- Dataset after Preprocessing ---


/tmp/ipykernel_2997/1141922053.py:7: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_processed['Age'].fillna(df_processed['Age'].median(), inplace=True)
/tmp/ipykernel_2997/1141922053.py:10: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(valu

,Survived,Pclass,Age,SibSp,Parch,Fare,Sex_male,Embarked_Q,Embarked_S
0,0,3,22.0,1,0,7.2500,True,False,True
1,1,1,38.0,1,0,71.2833,False,False,False
2,1,3,26.0,0,0,7.9250,False,False,True
3,1,1,35.0,1,0,53.1000,False,False,True
4,0,3,35.0,0,0,8.0500,True,False,True



--- Updated Dataset Info ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Survived    891 non-null    int64  
 1   Pclass      891 non-null    int64  
 2   Age         891 non-null    float64
 3   SibSp       891 non-null    int64  
 4   Parch       891 non-null    int64  
 5   Fare        891 non-null    float64
 6   Sex_male    891 non-null    bool   
 7   Embarked_Q  891 non-null    bool   
 8   Embarked_S  891 non-null    bool   
dtypes: bool(3), float64(2), int64(4)
memory usage: 44.5 KB

--- Missing Values after preprocessing ---
Survived      0
Pclass        0
Age           0
SibSp         0
Parch         0
Fare          0
Sex_male      0
Embarked_Q    0
Embarked_S    0
dtype: int64


## 4. Data Splitting

We will now split the preprocessed data into features (X) and the target variable (y). Then, we'll further divide these into training and testing sets to prepare for model building and evaluation.

In [8]:
from sklearn.model_selection import train_test_split

# Define features (X) and target (y)
X = df_processed.drop('Survived', axis=1)
y = df_processed['Survived']

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print("\n--- Dataset Split Information ---")
print(f"X_train shape: {X_train.shape}")
print(f"X_test shape: {X_test.shape}")
print(f"y_train shape: {y_train.shape}")
print(f"y_test shape: {y_test.shape}")


--- Dataset Split Information ---
X_train shape: (712, 8)
X_test shape: (179, 8)
y_train shape: (712,)
y_test shape: (179,)


## 5. Model Training and Evaluation

We will train a Logistic Regression model and then evaluate its accuracy on the test set to see how well it predicts survival.

In [9]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

# Initialize and train the Logistic Regression model
# Using 'liblinear' solver which is good for small datasets and handles L1/L2 penalties
model = LogisticRegression(random_state=42, solver='liblinear')
model.fit(X_train, y_train)

# Make predictions on the test set
y_pred = model.predict(X_test)

# Evaluate the model's accuracy
accuracy = accuracy_score(y_test, y_pred)
print(f"\nModel Accuracy: {accuracy:.4f}")


Model Accuracy: 0.7821


## 6. Conclusion and Next Steps

We have successfully built a Logistic Regression model to predict Titanic survival, achieving an accuracy of approximately 78.21%. This demonstrates a foundational understanding of the dataset and the application of a common machine learning algorithm.

### Potential Future Improvements:

*   **Feature Engineering:** Create new features from existing ones (e.g., family size from `SibSp` and `Parch`, title from `Name`).
*   **More Advanced Imputation:** Explore more sophisticated imputation techniques for 'Age' (e.g., using regression).
*   **Categorical Feature Handling:** Experiment with different encoding strategies for categorical variables.
*   **Model Selection:** Try other classification algorithms (e.g., Decision Trees, Random Forests, Support Vector Machines, Gradient Boosting).
*   **Hyperparameter Tuning:** Optimize the hyperparameters of the chosen model to improve performance.
*   **Cross-Validation:** Implement cross-validation for more robust model evaluation.
*   **Ensemble Methods:** Combine multiple models to potentially achieve higher accuracy.